### PR TITLE
snap: improve the error message for `snap try`

### PR DESCRIPTION
**Pull request by mvo5**
LP: #1593450

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1112,12 +1112,19 @@ func trySnap(c *Command, r *http.Request, user *auth.UserState, trydir string, f
 	if !filepath.IsAbs(trydir) {
 		return BadRequest("cannot try %q: need an absolute path", trydir)
 	}
-	if !osutil.IsDirectory(trydir) {
```
Comment (niemeyer) [inline daemon/api.go]: Should we really remove this? Note that the new error is in a branch that we only get into when it *is* a directory, while this check would only be triggered when it was *not* a directory. Does that mean we're not properly handling the case where try is handed a non-dir path?

File: daemon/api.go [hunk 2]
```diff
@@ -1112,12 +1112,19 @@ func trySnap(c *Command, r *http.Request, user *auth.UserState, trydir string, f
...
-		return BadRequest("cannot try %q: not a snap directory", trydir)
-	}
 
 	// the developer asked us to do this with a trusted snap dir
 	info, err := unsafeReadSnapInfo(trydir)
+	if _, ok := err.(snap.NoSnapDirError); ok {
```
Comment (niemeyer) [inline daemon/api.go]: This feels like a missing snap directory, when we supposedly were handed a snap directory and seem to be unhappy about it. How about.. hmmm.. `snap.NotSnapError`?  This way we can use it every time we're sure the content we've been provided is not even a snap (dir or file). We shouldn't use this when we are handed a snap, but its metadata is invalid (`snap.InvalidSnapError` would be better for this).

File: daemon/api.go [hunk 3]
```diff
@@ -1112,12 +1112,19 @@ func trySnap(c *Command, r *http.Request, user *auth.UserState, trydir string, f
...
+		return SyncResponse(&resp{
+			Type: ResponseTypeError,
+			Result: &errorResult{
+				Message: err.Error(),
+				Kind:    errorKindNoSnapDir,
```
Comment (niemeyer) [inline daemon/api.go]: Same for the kind: `errorKindNotSnap`?

File: cmd/snap/cmd_snap_op.go [hunk 4]
```diff
@@ -680,6 +680,11 @@ func (x *cmdTry) Execute([]string) error {
 	}
 
 	changeID, err := cli.Try(path, opts)
+	if e, ok := err.(*client.Error); ok && e.Kind == client.ErrorKindNoSnapDir {
+		return fmt.Errorf(i18n.G(`%q does not contain an unpacked snap.
+
+Try "snapcraft prime" in your project directory, then "snap try prime"`), path)
```
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: `then "snap try prime"` => `then "snap try" again.`

We seem to handle the prime case fine, per logic right above. Bonus points for the "try again" pun. ;D

File: daemon/api.go (hunk 1 above)
Comment (mvo5) [inline daemon/api.go]: Thanks, you are right, I restored it.

File: daemon/api.go (hunk 2 above)
Comment (mvo5) [inline daemon/api.go]: Thanks, I renamed it to `NotSnapError`

File: daemon/api.go (hunk 3 above)
Comment (mvo5) [inline daemon/api.go]: This one is renamed now as well

File: cmd/snap/cmd_snap_op.go (hunk 4 above)
Comment (mvo5) [inline cmd/snap/cmd_snap_op.go]: heh :) I like that! Updated.

File: client/client.go [hunk 5]
```diff
@@ -324,6 +324,8 @@ const (
 	ErrorKindSnapAlreadyInstalled = "snap-already-installed"
 	ErrorKindSnapNotInstalled     = "snap-not-installed"
 	ErrorKindNoUpdateAvailable    = "snap-no-update-available"
+
+	ErrorKindNotSnap = "snap-not-snap"
```
Comment (pedronis) [inline client/client.go]: I would use "snap-not-a-snap"  as the tag, and maybe even the kind but no strong opinion on the kind

Comment (mvo5) [inline client/client.go]: Indeed, I think that clearer. I fixed the tag. I'm a bit on the fence about the kind, i.e. `ErrorKindNotASnap` reads a bit strange (but maybe that is just me).

**Pull request merged**